doc/manpages/peldd.1
.TH PELDD 1
.SH NAME
peldd - shows library dependencies for a given PE file

.SH SYNOPSIS
.B peldd
[OPTIONS]...
.IR pefile

.SH DESCRIPTION
peldd shows library dependencies for a given PE file. It's part of pev, the PE file analysis toolkit.
.PP
\&\fIpefile\fR is a PE32/PE32+ executable or dynamic linked library file.

.SH OPTIONS
.TP
.BR \-f ", " \-\-format\ <text|csv|xml|html>
Change output format (default is text).

.TP
.BR \-v ", " \-\-verbose
Show more information about found items.

.TP
.BR \-V ", " \-\-version
Show version.

.TP
.BR \-\-help
Show help.

.SH EXAMPLES
Show library dependencies for \fBputty.exe\fP:
.IP
$ peldd putty.exe

.SH REPORTING BUGS
Please, check the latest development code and report at https://github.com/mentebinaria/readpe/issues

.SH SEE ALSO
\fBofs2rva\fP(1), \fBpedis\fP(1), \fBpehash\fP(1), \fBpepack\fP(1), \fBperes\fP(1), \fBpescan\fP(1), \fBpesec\fP(1), \fBpestr\fP(1), \fBreadpe\fP(1), \fBrva2ofs\fP(1)

.SH COPYRIGHT
Copyright (C) 2018 - 2020 pev authors. License GPLv2+: GNU GPL version 2 or later <https://www.gnu.org/licenses/gpl-2.0.txt>.
This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.


CHANGELOG
pev - PE32 Version utility changelog

Legend:
+ Added feature
* Improved/changed feature
- Bug fixed
! Known issue / missing feature

pev 0.40 - August 7, 2011
 + PE32+ support. Now pev can handle 64-bits executables.
 + Variable data directories support (no more fixed in 16).
 + Added more human-readable fields, like subsystem and section characteristics.
 + Added TLS callback functions detection in every PE section.
 + Added "-r" option to show resource items at first level.
 + ASLR and DEP identification.
 * Improved function to get machine type (thanks to Gabriel Duarte).
 * Date format in COFF header similar to RFC 2822.
 * Compatible with PE/COFF specification v8.2.

pev 0.31 - May 11, 2011
 + Added human-readable machine types in COFF output.
 + Added characteristics flags in COFF output.
 - Fixed warning with Linux 32-bits boxes when compiling.
 - Now pev compiles in OSX without any changes (thanks to Gustavo Roberto).

pev 0.30 - February 20, 2011
 + Added option "-d" to show the DOS header.
 + Added option "-c" to show the COFF header.
 + Added option "-o" to show the Option (PE) header.
 + Added option "-s" to show executable sections.
 + Added option "-a" to show all information.
 * Improved memory use.
 * Now pev shows the Product Version with option "-p".
 
pev 0.22 - January 9, 2011
 + Added manpage.
 * Improved Makefile.
 ! Plans to read more PE informations.
 ! Does not support PE32+ files.

pev 0.2 - December 26, 2011
 * Improved search algorithm.
 - Fixed bug compiling in MS-Windows platform.

pev 0.1 - December 12, 2010
   Initial release.


src/peldd.c
/* vim: set ts=4 sw=4 noet: */
/*
	pev - the PE file analyzer toolkit

	peldd.c - shows library dependencies for a given PE file

	Copyright (C) 2018 - 2020 pev authors

	This program is free software: you can redistribute it and/or modify
	it under the terms of the GNU General Public License as published by
	the Free Software Foundation, either version 2 of the License, or
	(at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program.  If not, see <http://www.gnu.org/licenses/>.

	In addition, as a special exception, the copyright holders give
	permission to link the code of portions of this program with the
	OpenSSL library under certain conditions as described in each
	individual source file, and distribute linked combinations
	including the two.

	You must obey the GNU General Public License in all respects
	for all of the code used other than OpenSSL.  If you modify
	file(s) with this exception, you may extend this exception to your
	version of the file(s), but you are not obligated to do so.  If you
	do not wish to do so, delete this exception statement from your
	version.  If you delete this exception statement from all source
	files in the program, then also delete it here.
*/

#include "peldd.h"
#include "main.h"
#include "common.h"
#include "output.h"
#include <libpe/macros.h>
#include <ctype.h>

#define PROGRAM "peldd"

static void usage(void)
{
	static char formats[255];
	output_available_formats(formats, sizeof(formats), '|');
	printf("Usage: %s FILE\n"
		"Display PE library dependencies\n"
		"\nExample: %s winzip.exe\n"
		"\nOptions:\n"
		" -f, --format <%s>  Change output format (default: text).\n"
		" -V, --version							 Show version.\n"
		" --help								 Show help.\n",
		PROGRAM, PROGRAM, formats);
}

static void parse_options(int argc, char *argv[])
{
	/* Parameters for getopt_long() function */
	static const char short_options[] = "Vf:";

	static const struct option long_options[] = {
		{ "help",			  no_argument,		 NULL,	1  },
		{ "format",			  required_argument, NULL, 'f' },
		{ "version",		  no_argument,		 NULL, 'V' },
		{  NULL,			  0,				 NULL,	0  }
	};

	int c, ind;

	while ((c = getopt_long(argc, argv, short_options, long_options, &ind)))
	{
		if (c < 0)
			break;

		switch (c)
		{
			case 1: // --help option
				usage();
				exit(EXIT_SUCCESS);
			case 'V':
				printf("%s %s\n%s\n", PROGRAM, TOOLKIT, COPY);
				exit(EXIT_SUCCESS);
			case 'f':
				if (output_set_format_by_name(optarg) < 0)
					EXIT_ERROR("invalid format option");
				break;
			default:
				fprintf(stderr, "%s: try '--help' for more information\n", PROGRAM);
				exit(EXIT_FAILURE);
		}
	}
}

void print_dependencies(pe_ctx_t *ctx)
{
	output_open_scope("Dependencies", OUTPUT_SCOPE_TYPE_ARRAY);
	const pe_imports_t *imports = pe_imports(ctx);
	for (size_t i=0; i < imports->dll_count; i++) {
		const pe_imported_dll_t *dll = &imports->dlls[i];
		output(dll->name, NULL);
	}
	output_close_scope();
}

int peldd(int argc, char *argv[])
{
	pev_config_t config;
	PEV_INITIALIZE(&config);

	if (argc < 2) {
		usage();
		return EXIT_FAILURE;
	}

	output_set_cmdline(argc, argv);

	parse_options(argc, argv);

	pe_ctx_t ctx;

	pe_err_e err = pe_load_file(&ctx, argv[argc-1]);
	if (err != LIBPE_E_OK) {
		pe_error_print(stderr, err);
		return EXIT_FAILURE;
	}

	err = pe_parse(&ctx);
	if (err != LIBPE_E_OK) {
		pe_error_print(stderr, err);
		return EXIT_FAILURE;
	}

	if (!pe_is_pe(&ctx))
		EXIT_ERROR("not a valid PE file");

	output_open_document();

	IMAGE_DATA_DIRECTORY **directories = pe_directories(&ctx);
	if (directories == NULL) {
		LIBPE_WARNING("directories not found");
	} else {
		print_dependencies(&ctx);
	}

	output_close_document();

	// free
	err = pe_unload(&ctx);
	if (err != LIBPE_E_OK) {
		pe_error_print(stderr, err);
		return EXIT_FAILURE;
	}

	PEV_FINALIZE(&config);

	return EXIT_SUCCESS;
}
